File: pe/submissions/accepted/fenwick.py
import sys

def main():
    data = sys.stdin.read().split()
    n = int(data[0])
    v = [0] * (n + 2)
    bit = [0] * (n + 2)

    idx = 1
    for i in range(1, n + 1):
        x = int(data[idx])
        v[x] = i
        idx += 1

    ans = n * (n - 1) // 2

    for i in range(n):
        x = int(data[idx])
        idx += 1
        pos = v[x]
        # query
        res = 0
        j = pos
        while j:
            res += bit[j]
            j -= j & -j
        ans -= res
        # update
        j = pos
        while j <= n:
            bit[j] += 1
            j += j & -j

    print(ans)

if __name__ == '__main__':
    main()

#CPU time: 0.969 seconds
#Peak memory usage: 165.02 MiB

File: pe/submissions/accepted/fenwick.c
#include <stdio.h>
#define N 1000000
#define lsb(x) 1 << __builtin_ctz(x)

int n, bit[N + 1];
void update(int i) {
    for (; i <= n; i += lsb(i))
        ++bit[i];
}

int query(int i) {
    int ans = 0;
    for (; i; i ^= lsb(i))
        ans += bit[i];

    return ans;
}

int v[N + 1];
int main(void) {
    scanf("%d", &n);

    int i, x;
    for (i = 1; i <= n; ++i) {
        scanf("%d", &x);
        v[x] = i;
    }

    long long ans = (1ll * n * (n - 1)) >> 1ll;
    for (i = 1; i <= n; ++i) {
        scanf("%d", &x);
        ans -= query(v[x]);
        update(v[x]);
    }

    printf("%lld\n", ans);
    return 0;
}
